### ISSUE TITLE
Design and implement post-rendering validation

### ISSUE THREAD

**Issue by drevell** [abcxyz/abc#35]
### TL;DR

As a template developer, I want to be able to add checks after template rendering to make sure the resulting code is correct. If these validations fail, I want the CLI to show an error message to the user telling them that the template spec file that they're is buggy, and ask them to file a bug to the template owner.

Ideas for use cases:
 - Assert that the resulting code can build and compile
 - Assert that a given string or regex does or doesn't appear. For example, "template_replace_me" shouldn't appear because it should be replaced in rendering.
 - Assert that unit tests pass
 - EDIT: maybe also `go mod tidy`

One possible design: the template spec.yaml file could declare a list of validations, like:
```
- validations:
  - validator: 'go_lint'  
  - validator: 'go_compile_and_test'
  - validator: 'strings_must_not_appear'
    strings:
      - 'template_replace_me'
```

**Comment by drevell** [comment]
Meeting note: I'm going to work on this, and @chloechien will contribute npm support. We will want a design doc.

**Comment by yolocs** [comment]
How to validate a template could vary a lot. How about we leave it to the template repo GH workflow to do? My rough thoughts are:

I'm in my template repo. As a presubmit check, it would:

1. Use the abc CLI (either `latest` or `head`) to render the result into a folder
2. From this folder I could do various checks
   - The basic one: check if the content matches my "want" version - the CLI could provide this functionality since it's so common
   - Run lint and unit tests - the CLI doesn't need to worry about this; I get to choose what tools to use to run my test or lint
   - All kinds of other checks if needed
